### PR TITLE
Handling HTTP Cookies properly in SAML authentication

### PR DESCRIPTION
**Pull request by ruwanta**

### CODE REVIEW

File: components/org.wso2.carbon.appmgt.gateway/src/main/java/org/wso2/carbon/appmgt/gateway/handlers/common/HttpCookieUtil.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            List<HttpCookie> setCookies) {
+        Map<String, Object> headers = (Map<String, Object>) axis2MC
+                .getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+        Map<String, Object> excessHeaders = (Map<String, Object>) axis2MC
+                .getProperty(NhttpConstants.EXCESS_TRANSPORT_HEADERS);
+
+        if (setCookies.size() > 0) {
+            headers.put(HTTPConstants.HEADER_SET_COOKIE, HttpCookieUtil.formatSetCookieHeader((setCookies.get(0))));
+            if (setCookies.size() > 1) {
+                excessHeaders.remove(HTTPConstants.HEADER_SET_COOKIE);
```
Comment (ruwanta) [inline components/org.wso2.carbon.appmgt.gateway/src/main/java/org/wso2/carbon/appmgt/gateway/handlers/common/HttpCookieUtil.java]: Remove from cookies from excess headers need to be done unconditionally

File: components/org.wso2.carbon.appmgt.gateway/src/main/java/org/wso2/carbon/appmgt/gateway/handlers/security/saml2/SAML2AuthenticationHandler.java [hunk 2]
```diff
@@ -51,29 +48,8 @@
...
-import org.opensaml.saml2.core.SessionIndex;
-import org.opensaml.saml2.core.impl.AuthnContextClassRefBuilder;
-import org.opensaml.saml2.core.impl.AuthnRequestBuilder;
-import org.opensaml.saml2.core.impl.IssuerBuilder;
-import org.opensaml.saml2.core.impl.LogoutRequestBuilder;
-import org.opensaml.saml2.core.impl.NameIDBuilder;
-import org.opensaml.saml2.core.impl.NameIDPolicyBuilder;
-import org.opensaml.saml2.core.impl.RequestedAuthnContextBuilder;
-import org.opensaml.saml2.core.impl.SessionIndexBuilder;
+import org.opensaml.saml2.core.*;
```
Comment (ruwanta) [inline components/org.wso2.carbon.appmgt.gateway/src/main/java/org/wso2/carbon/appmgt/gateway/handlers/security/saml2/SAML2AuthenticationHandler.java]: Import expansion

File: components/org.wso2.carbon.appmgt.gateway/src/main/java/org/wso2/carbon/appmgt/gateway/handlers/security/saml2/SAML2AuthenticationHandler.java [hunk 3]
```diff
@@ -437,6 +374,37 @@ public boolean handleResponse(MessageContext messageContext) {
...
+        samleSsoCookie.setHttpOnly(true);
+        samleSsoCookie.setPath("/");
+        if (log.isDebugEnabled()) {
+            log.debug("Updated set cookie string in transport headers : " + appmSamlSsoCookie);
+        }
+
+        HttpCookieUtil.writeSetCookieHeaders(axis2MC, allSetCookies);
+    }
+
+    private HttpCookie parseSetCookieHeader(String cookieString) {
```
Comment (ruwanta) [inline components/org.wso2.carbon.appmgt.gateway/src/main/java/org/wso2/carbon/appmgt/gateway/handlers/security/saml2/SAML2AuthenticationHandler.java]: Can use direct call to Utility class

File: components/org.wso2.carbon.appmgt.gateway/src/main/java/org/wso2/carbon/appmgt/gateway/handlers/security/saml2/SAML2AuthenticationHandler.java [hunk 4]
```diff
@@ -1753,33 +1721,34 @@ private String constructAssertionConsumerUrl(MessageContext messageContext) {
...
 
-        if (log.isDebugEnabled()) {
-            log.debug("Exisiting cookie string in transport headers : " + cookieString);
-        }
-        if (cookieString == null) {
-            cookieString = AppMConstants.APPM_SAML2_COOKIE + "=" + appmSamlSsoCookie + "; " + "path=" + "/";
-        } else {
-            cookieString = cookieString + " ;" + AppMConstants.APPM_SAML2_COOKIE + "=" + appmSamlSsoCookie + ";" + " Path=" + "/";
+        //Search for the cookie containing APPM_SAML2_COOKIE
+        HttpCookie samleSsoCookie = null;
```
Comment (ruwanta) [inline components/org.wso2.carbon.appmgt.gateway/src/main/java/org/wso2/carbon/appmgt/gateway/handlers/security/saml2/SAML2AuthenticationHandler.java:1760]: samleSsoCookie -> samlSsoCookie

**Comment by isharak** [comment]
Please resolve the conflicts and resend the PR

**Pull request closed without merging**